### ISSUE TITLE
Complex `linalg.pinv` / `linalg.polar` / `linalg.matrix_sqrth` return wrong values under any `TorchDispatchMode` (hence under `torch.compile`): a conjugate view taken inside the ATen kernel is not applied

### ISSUE THREAD

**Issue by cuiliaomei-beep** [pytorch/pytorch#198118]
### 🐛 Describe the bug

**Summary.** Complex `torch.linalg.pinv`, `linalg.polar` and `linalg.matrix_sqrth` return one result in plain eager and another while any `TorchDispatchMode` is active, even one that only calls `func(*args)`, and `torch.compile` returns the second because its runtime runs every graph under `_AnalyzeCustomOpInputOutputMode` (torch 2.14.0 and nightly; float inputs are unaffected). A user gets `A @ pinv(A) @ A != A` and `U @ P != A` with no error, and every gradient built on these values (#197084) is wrong for the same reason. Because the kernel conjugates the SVD factors internally with `.mH()` and that lazy conjugate view is consumed unmaterialized when the `Python` dispatch key is in the thread-local include set, rather than through the Conjugate fallback that the same formula written op by op uses, the result equals `vh.mT @ diag(1/s) @ u.mH`.

```python
import torch
from torch.utils._python_dispatch import TorchDispatchMode

class Forward(TorchDispatchMode):            # does nothing but forward
    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        return func(*args, **(kwargs or {}))

def manual_pinv(a):                          # the same formula, op by op
    u, s, vh = torch.linalg.svd(a, full_matrices=False)
    return vh.mH @ torch.diag(1 / s).to(a.dtype) @ u.mH

torch.manual_seed(0)
A = torch.randn(4, 3, dtype=torch.complex64)
ref = torch.linalg.pinv(A)
with Forward():
    under_mode, manual = torch.linalg.pinv(A), manual_pinv(A)
torch._dynamo.reset()
compiled = torch.compile(torch.linalg.pinv)(A)
u, s, vh = torch.linalg.svd(A, full_matrices=False)
print("|pinv under mode - eager|   ", (under_mode - ref).abs().max().item())
print("|pinv compiled   - eager|   ", (compiled - ref).abs().max().item())
print("|op-by-op under mode - eager|", (manual - ref).abs().max().item())
print("under mode == vh.mT @ S^-1 @ u.mH:", torch.allclose(under_mode, vh.mT @ torch.diag(1 / s).to(A.dtype) @ u.mH, atol=1e-5))
print("A @ pinv(A) @ A == A:  eager", torch.allclose(A @ ref @ A, A, atol=1e-5), " compiled", torch.allclose(A @ compiled @ A, A, atol=1e-5))
P = torch.linalg.polar(A)[1]
with Forward():
    P_mode = torch.linalg.polar(A)[1]
print("polar P[0, 1]: eager", P[0, 1].item(), " under mode", P_mode[0, 1].item())
```

```
|pinv under mode - eager|    1.1541486978530884
|pinv compiled   - eager|    1.1541486978530884
|op-by-op under mode - eager| 0.0
under mode == vh.mT @ S^-1 @ u.mH: True
A @ pinv(A) @ A == A:  eager True  compiled False
polar P[0, 1]: eager (0.34345024824142456+0.11318771541118622j)  under mode (0.34345024824142456+0j)
```

How this was narrowed down: the compiled graph contains only `aten.linalg_polar.default` / `aten.linalg_pinv.atol_rtol_tensor` (no decomposition), and running that graph module directly gives the eager result; inside AOTAutograd's runtime wrapper the same call is wrong, and the only difference in dispatcher state there is the active `_AnalyzeCustomOpInputOutputMode` (TLS include set gains `Python` and `PythonTLSSnapshot`). A plain forwarding mode reproduces it, and so does eager under `torch._C._ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Conjugate))`, which gives the identical wrong tensors. Scanning every `torch.linalg` function and 51 complex composite calls under the forwarding mode, the affected ones are `linalg.pinv` / `pinverse`, `linalg.polar` (the Hermitian factor `P`, which comes back as `(X + X.mT) / 2` so the imaginary parts of its off-diagonal cancel) and `linalg.matrix_sqrth`; `lstsq`, `solve`, `inv`, `eigh`, `cholesky_*`, `lu_solve(adjoint=True)`, `householder_product`, `ormqr`, `vdot`, `matmul` with `.mH` and the rest agree.

This is also the actual cause of #197084 (wrong gradients of complex `pinv` / `matrix_sqrth` under `torch.compile`): with the forwarding mode active in eager, the gradient of `(pinv(a).abs() ** 2).sum()` differs from the double-precision reference by the same `1.7e+01` and is bit-identical to the compiled gradient, and it stays wrong when only the forward runs under the mode, so the backward formula is fine and it is the saved forward output that is wrong.

**Possible fix.** Inside these kernels the conjugate views produced by `.mH()` are consumed by `at::matmul` on the path taken when the `Python` key is in the TLS include set; either these kernels should call `.resolve_conj()` (or `conj_physical`) before the matmul, or the Python fallback should preserve the Conjugate/Negative fallbacks when it re-dispatches for ops that were not intercepted by a mode. A regression test that runs the complex `linalg` OpInfo samples under a forwarding `TorchDispatchMode` would catch this class.

### Versions

```
PyTorch version: 2.14.0+cpu
OS: Microsoft Windows 11 Pro for Workstations (10.0.26200 64-bit)
Python version: 3.14.7 (64-bit runtime)
Is CUDA available: False
CPU: 11th Gen Intel(R) Core(TM) i9-11900T @ 1.50GHz
[pip3] numpy==2.5.2
[pip3] torch==2.14.0+cpu
```

Same output on nightly 2.15.0.dev20260911+cpu and on Linux (2.14.0+cpu, nightly 2.15.0.dev20260921+cpu). On a Tesla T4 (2.14.0+cu130 and nightly) `linalg.polar` is wrong in the same way (`|P under mode - eager| = 2.07` for complex64), while `linalg.pinv` on CUDA agrees with eager, so the CUDA `pinv` kernel does not take the affected path.

cc @ezyang @anjali411 @dylanbespalko @mruberry @nikitaved @amjames @jianyuh @walterddr @Lezcano @Chillee @albanD @samdow

**Comment by cuiliaomei-beep** [comment]
Two additions after sweeping the gradients as well (every OpInfo op with autograd, complex64/complex128/float64, eager vs the same computation under a forwarding `TorchDispatchMode`; 955 operator–dtype pairs):

1. **The backward of `torch.linalg.lstsq` is affected too.** Its forward is identical under the mode, but the gradient is not — the C++ `linalg_lstsq_backward` goes through `pinv` and `.mH()` internally. Apart from `lstsq`, only the gradients of `pinv` / `pinverse` and `matrix_sqrth` change (their forwards are the ones in the description); all float64 gradients are unchanged. `torch.compile` happens **not** to be affected for `lstsq`, because AOTAutograd traces the derivative formula op by op instead of calling the composite C++ backward under the mode.

2. **A stock PyTorch mode triggers it**, so this is not specific to `torch.compile`: `torch.utils.flop_counter.FlopCounterMode` — used to profile a training step — gives the same wrong values.

```python
import torch
from torch.utils.flop_counter import FlopCounterMode

torch.manual_seed(0)
A = torch.randn(5, 3, dtype=torch.complex64); B = torch.randn(5, 2, dtype=torch.complex64)

def loss(A, B):
    return (torch.linalg.lstsq(A, B).solution.abs() ** 2).sum()

def grad(A, B, mode=None):
    A = A.detach().clone().requires_grad_(); B = B.detach().clone().requires_grad_()
    if mode is None:
        loss(A, B).backward()
    else:
        with mode:
            loss(A, B).backward()
    return A.grad

ref = grad(A.to(torch.complex128), B.to(torch.complex128))
print("lstsq grad  eager |dA - ref|", (grad(A, B).to(torch.complex128) - ref).abs().max().item())
print("lstsq grad  under FlopCounterMode |dA - ref|", (grad(A, B, FlopCounterMode(display=False)).to(torch.complex128) - ref).abs().max().item())
M = torch.randn(4, 3, dtype=torch.complex64)
with FlopCounterMode(display=False):
    p = torch.linalg.pinv(M)
print("pinv        under FlopCounterMode |pinv - eager|", (p - torch.linalg.pinv(M)).abs().max().item())
```

```
lstsq grad  eager |dA - ref| 4.480225896255682e-05
lstsq grad  under FlopCounterMode |dA - ref| 173.15849216723038
pinv        under FlopCounterMode |pinv - eager| 1.285477638244629
```

So any code that profiles or logs a complex-valued model with a dispatch mode (FLOP counting, logging tensors, memory tracking) trains on wrong gradients for these ops, in plain eager. The fix location is the same: the conjugate views created inside the ATen composites (`linalg_pinv`, `linalg_polar`, `linalg_matrix_sqrth`, and `linalg_lstsq_backward`) are consumed without being materialized while the `Python` dispatch key is in the TLS include set.
